third_party/xla_client/thread_pool.cc

```diff
@@ -1,18 +1,180 @@
+        break;
+      }
+      try {
+        closure();
+      } catch (const std::exception& ex) {
+        XLA_COUNTER("ThreadPoolException", 1);
+        TF_LOG(ERROR) << "Exception from running thread pool closure: "
+                      << ex.what();
+      }
+    }
+  }
+
+  void ScheduleOnThread(std::function<void()> closure) {
+    std::thread thread(std::move(closure));
+    thread.detach();
+  }
+
+  std::function<void()> GetWork() {
+    std::unique_lock<std::mutex> lock(mutex_);
+    ++waiting_;
+    cv_.wait(lock, [this] { return done_ || !work_.empty(); });
+    --waiting_;
+    if (work_.empty()) {
+      return nullptr;
+    }
+    std::function<void()> closure(std::move(work_.front()));
+    work_.pop_front();
+    return closure;
+  }
+
+  std::vector<std::thread> threads_;
+  std::mutex mutex_;
+  std::condition_variable cv_;
+  bool done_ = false;
+  std::deque<std::function<void()>> work_;
+  size_t waiting_ = 0;
+};
+
+ThreadPool* GetThreadPool() {
+  static size_t num_threads = sys_util::GetEnvInt(
+      "XLA_THREAD_POOL_SIZE", std::thread::hardware_concurrency());
+  static ThreadPool* pool = new ThreadPool(num_threads);
+  return pool;
+}
+
+ThreadPool* GetIoThreadPool() {
+  static size_t num_threads = sys_util::GetEnvInt(
+      "XLA_IO_THREAD_POOL_SIZE", std::thread::hardware_concurrency());
+  static ThreadPool* pool = new ThreadPool(num_threads);
+  return pool;
+}
+
+}  // namespace
+
+class Completion::Data {
+ public:
+  void Wait() {
+    std::unique_lock<std::mutex> lock(mutex_);
+    cv_.wait(lock, [this] { return completed_; });
+    if (exptr_ != nullptr) {
+      std::rethrow_exception(exptr_);
+    }
+  }
+
+  static std::function<void()> GetCompleter(std::shared_ptr<Data> data,
+                                            std::function<void()> closure) {
+    auto closure_wrapper = [closure = std::move(closure), data]() {
+      std::exception_ptr exptr;
+      try {
+        closure();
+      } catch (...) {
+        exptr = std::current_exception();
+      }
+      data->Complete(exptr);
+    };
+    return closure_wrapper;
+  }
+
+ private:
+  void Complete(std::exception_ptr exptr) {
+    std::lock_guard<std::mutex> lock(mutex_);
+    exptr_ = std::move(exptr);
+    completed_ = true;
+    cv_.notify_all();
+  }
+
+  std::mutex mutex_;
+  std::condition_variable cv_;
+  bool completed_ = false;
+  std::exception_ptr exptr_;
+};
+
+Completion::Completion(std::shared_ptr<Data> data) : data_(std::move(data)) {}
+
+Completion::~Completion() {}
+
+void Completion::Wait() { data_->Wait(); }
 
 void ScheduleClosure(std::function<void()> closure) {
-  std::thread t(std::move(closure));
-  t.detach();
+  GetThreadPool()->Schedule(std::move(closure));
 }
 
 void ScheduleIoClosure(std::function<void()> closure) {
-  std::thread t(std::move(closure));
-  t.detach();
+  GetIoThreadPool()->Schedule(std::move(closure));
+}
+
+Completion ScheduleClosureWithCompletion(std::function<void()> closure) {
```

Comment:
We don't use these two methods, do you plan to use them for something in the near future?

Reply:
Yes, the main reason I did this, is because there is a chance of sending part of the get-tensors in background.
But we need something to wait for, because we need to do extra work after that.